design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+
+### Admin operations
+
+#### `SHOW TOPICS`
+
+```
+  CompletableFuture<List<TopicInfo>> listTopics();
+```
+with
+```
+public interface TopicInfo {
+
+  String getName();
+
+  int getPartitions();
+  
+  List<Integer> getReplicasPerPartition();
+
+}
+```
+
+#### `SHOW <STREAMS/TABLES>`
+
+```
+  CompletableFuture<List<StreamInfo>> listStreams();
+
+  CompletableFuture<List<TableInfo>> listTables();
+```
+with
+```
+public interface StreamInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+  
+}
+```
+and
+```
+public interface TableInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+
+  boolean isWindowed();
+  
+}
+```
+
+I'm not sure whether it makes more sense for `StreamInfo#getFormat()` and `TableInfo#getFormat()` to return a string or an enum value.
+The latter would make it easier for the user to know the possible values, but we'd have to keep the list up to date and would also sacrifice forward compatibility.
+
+#### `SHOW QUERIES`
+
+```
+  CompletableFuture<List<QueryInfo>> listQueries();
+```
+with
+```
+public interface QueryInfo {
+
+  boolean isPersistentQuery();
+
+  boolean isPushQuery();
+
+  /**
+   * Query ID, used for control operations such as terminating the query
+   */
+  String getId();
+
+  String getSql();
+
+  /**
+   * Name of sink, for a persistent query. Else, null.
+   */
+  String getSink();
+
+  /**
+   * Name of sink topic, for a persistent query. Else, null.
+   */
+  String getSinkTopic();
+  
+  /**
+   * Map of query state ("RUNNING", "ERROR", or "UNRESPONSIVE") to the number of ksqlDB servers
+   * on which this query is in each state.
+   */
+  Map<String, Integer> getServerStatusCounts();
+
+}
+```
+
+Again, not sure whether it makes more sense to represent query states as strings, or to introduce an Enum and have `getServerStatusCounts()` return an EnumMap instead.
+
+Does having `getSink()` and `getSinkTopic()` return null in the case of push queries make sense, or would we prefer an empty string or an Optional instead?
+
+#### `SHOW CONNECTORS`
+
+```
+  CompletableFuture<ConnectorList> listConnectors();
+```
+with
+```
+public interface ConnectorList {
+  
+  List<ConnectorInfo> getConnectors();
+  
+  /**
+   * Any warnings returned by the server as a result of listing connectors.
+   */
+  List<String> getWarnings();
+
+}
+```
+and
+```
+public interface ConnectorInfo {
+
+  enum ConnectorType {
+    SOURCE,
+    SINK,
+    UNKNOWN;
+  }
+
+  String getName();
+
+  ConnectorType getType();
+
+  String getClassName();
+
+  String getState();
+
+}
+```
+
+I don't love that the introduction of `ConnectorList` wrapped around `List<ConnectorInfo>` breaks the pattern established by `SHOW TOPICS`/`SHOW <STREAMS/TABLES>`/`SHOW QUERIES` but it seems important to propagate any server warnings to the user so the trade-off is worth it IMO.
+
+### Terminate push query
+
+```
+  CompletableFuture<Void> terminatePushQuery(String queryId);
```

Comment:
how do you get the queryId for a push query? (related to my comment above about how do you terminate push queries)

Reply:
It's returned as part of the `QueryResult` from `streamQuery(...)`

Reply:
Why do we need method for that? It seem to be much simpler, to just say (just to sketch the idea):
```
QueryHandle handle = client.streamQuery(...):
handle.terminate();
```

This way, we don't even need to expose the queryId at all.

Reply:
The thing that's closing the query might not have access to the QueryHandle. It might be in a different process, on a different machine. Much easier to pass strings around.

Reply:
Why would this be the case? If I issue a query, I would not allow anybody else to terminate my query. Can you elaborate on the use case?

Reply:
I'm sure there are lots of cases. Maybe you have a distributed system where queries are created by one microservice, and you have another one that cleans up old queries, or closes them according to some criteria. These service could be on a completely different machine, or even written in a different language.

Reply:
Sound like premature optimization for an unknown use case. I would prefer to keep the API tight, and if a real need arises, we can still expose the query ID later and extend the API accordingly. Thoughts?

Reply:
Disagree that this is an optimisation. It's the simplest way I can think of doing it. Returning an object seems somewhat more work, and limits the usefulness.

Reply:
I guess this is subjective. I find dealing with "Strings" instead of proper object not simpler but clumsy. But I can only share/rise my opinion/preference.